File: 1972-rotating-the-box/rotating-the-box.py
class Solution:
    def rotateTheBox(self, box: List[List[str]]) -> List[List[str]]:
        m = len(box)
        n = len(box[0])

        # rotate 90 degree clockwise by transpose and reverse
        res = list(map(list, zip(*box[::-1])))

        new_rows = n
        new_cols = m

        # apply gravity
        for j in range(new_cols):
            lowest_empty_space = new_rows - 1
            for i in range(new_rows - 1, -1, -1):
                if res[i][j] == "*":
                    lowest_empty_space = i - 1
                elif res[i][j] == "#":
                    # move stone to lowest empty space
                    res[i][j] = "."
                    res[lowest_empty_space][j] = "#"
                    lowest_empty_space -= 1

        return res

File: 1972-rotating-the-box/rotating-the-box.cpp
class Solution {
public:
    vector<vector<char>> rotateTheBox(vector<vector<char>>& box) {
        int m = box.size();
        int n = box[0].size();

        vector<vector<char>> result(n, vector<char>(m, '.'));

        for (int i = 0; i < m; i++) {
            int lowest_empty_space = n - 1;
            for (int j = n - 1; j >= 0; j--) {
                if (box[i][j] == '#') {
                    result[lowest_empty_space][m - 1 - i] = '#';
                    lowest_empty_space--;
                } else if (box[i][j] == '*') {
                    result[j][m - 1 - i] = '*';
                    lowest_empty_space = j - 1;
                }
            }
        }
        return result;
    }
};